Make token controller resilient to resource conflicts

liggitt opened the pull request:
Includes changes from #8880
1. When removing service account references to a deleted token, retries on resource conflicts if the removed token is still referenced
2. Before auto-creating a token, make sure the controller is working with the latest version of the service account. If not, skip creation (which would fail and require cleanup) and wait for another update notification.
3. Added tests for resource conflict on creation

liggitt commented:
/cc @deads2k

liggitt commented:
/cc @erictune

erictune commented on pkg/serviceaccount/tokens_controller.go:
```diff
@@ -240,8 +244,16 @@ func (e *TokensController) secretDeleted(obj interface{}) {
 
-	if _, err := e.removeSecretReferenceIfNeeded(serviceAccount, secret.Name); err != nil {
-		glog.Error(err)
+	for i := 1; i <= NumServiceAccountRemoveReferenceRetries; i++ {
```
If the controller is stopped, and then a secret is deleted, and then the controller restarts, this code won't see the delete event.  

Therefore, to be really reliable, the controller needs to relist all ServiceAccounts upon startup and see if there are any that have dangling references to secrets, and removeSecretReferenceIfNeeded.

Given that it has to do this at startup, and that it need to be restartable, you might as well just run this check periodically.  Which obviates the need for the code below.

lavalamp replied:
Whenever a service account is processed, it should regenerate its list of secrets. Then I think you don't have to do this at all?

liggitt replied:
> Whenever a service account is processed, it should regenerate its list of secrets

Not sure what you mean by that... do you mean clean dangling references?

lavalamp replied:
I mean delete the current list; then search through all secrets and rebuild the list. Just do it every time you process the service account; it should be quick since you're using an indexing store.

liggitt replied:
The list can't be determined by inspecting secrets. The service account's list of secrets includes not only the token secrets, but any other secret you want pods referencing the service account to be able to mount.

lavalamp replied:
In that case, is it important to remove non-existing secrets from this list?

I feel like the water is very muddy here, in terms of what objects own/depend on other objects. Maybe after 1.0 we can try to simplify this; it's difficult to reason about at the moment.

liggitt replied:
> In that case, is it important to remove non-existing secrets from this list?

In this case, it's not critical, just bothers my referential-integrity-tuned sensibilities. I'm ok tabling this for the moment and revisiting

erictune commented on pkg/serviceaccount/tokens_controller.go:
```diff
@@ -270,6 +282,21 @@ func (e *TokensController) createSecretIfNeeded(serviceAccount *api.ServiceAccou
 // createSecret creates a secret of type ServiceAccountToken for the given ServiceAccount
 func (e *TokensController) createSecret(serviceAccount *api.ServiceAccount) error {
+	// We don't want to update the cache's copy of the service account
+	// so add the secret to a freshly retrieved copy of the service account
```
@lavalamp is this best practice?

lavalamp replied:
I would use deep copy here instead of re-getting from apiserver.

liggitt replied:
This was serving two purposes:
1. Getting from the apiserver gets us a copy we can mutate and update
2. More importantly, we also get to skip a create and cleanup if the event we're processing is outdated. If multiple things are adding secrets to service accounts, several update events can queue up in the process queue. Not doing this check results in a lot of thrash as different controllers attempt to respond to serviceaccount updates by ensuring their particular type of secret is generated, then added, only to get a resource conflict and have to clean up their secret and try again.

lavalamp replied:
If 2 is a concern, perhaps you should be using the work queue I added for endpoints controller.

liggitt replied:
I can look into that

liggitt committed:
Retry service account update when adding token reference

liggitt commented:
Moved this back to WIP while I look into the work queue options. Split out @deads2k's original token cleanup change into #9104 for cleanliness, since that needs doing anyway

deads2k commented:
unit tests are out of date.  Fix up "new serviceaccount with no secrets with unsynced secret store" and "updated serviceaccount with no secrets with unsynced secret store"

liggitt commented:
Ok, will take a look

deads2k commented:
@liggitt rebase

erictune commented:
I'm on vacation this week (:house: :mouse:) and will take a look next week.

erictune commented:
LGTM

deads2k commented:
@liggitt  rebase.  It'd be nice to stop carrying this.

liggitt commented:
rebased

dchen1107 commented:
triggered shippable again!

dchen1107 commented:
Also removed lgtm since the pr was updated after that lgtm.

liggitt commented:
I self added the LGTM since the "update" was a rebase conflict in the imports section. Trying to spare Eric review workload

liggitt commented:
@brendandburns not sure what happened here with the merge bot and the LGTM labels. In this case, there was no harm, but I'm wondering if there's a bug in how the bot checks LGTM labels before merging (or if I'm just misunderstanding its logic)
